### PR TITLE
Use lintrunner

### PR DESCRIPTION
**Pull request by justinchuby**
Adopt lintrunner for running all linters locally and in CI. https://github.com/suo/lintrunner

- Remove all formatting scripts in favor of lintrunner
- Updated readme with new instructions
- Run `lintrunner -a` to fix all lint errors
- Support SARIF results in Github

Nice SARIF error messages:

e.g. https://github.com/microsoft/onnx-script/security/code-scanning/335

image

### CODE REVIEW

**Comment by justinchuby** [comment]
@abock @gramalingam please take a look, thanks! I would like to make sure licensing and attribution is good for files under `.lintrunner`

File: pyproject.toml [hunk 1]
```diff
@@ -116,5 +116,6 @@ convention = "google"
 # D1 is for missing docstrings, which is not yet enforced.
 # D202 Too strict. "No blank lines allowed after function docstring"
 # D205 Too strict. "1 blank line required between summary line and description"
+# D401 Too strict. "First line should be in imperative mood"
```
Comment (AllenTiTaiWang) [inline pyproject.toml:119]: What was blocked by D401?

Comment (AllenTiTaiWang) [inline pyproject.toml:119]: What was blocked by D401?

File: .lintrunner.toml [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    'flake8==5.0.4',
+    'flake8-bugbear==22.10.27',
+    'flake8-pyi==22.10.0',
+    'dlint==0.13.0',
+    'toml==0.10.2',  # Needed for pydocstrings to parse pyproject.toml
+    'flake8-docstrings==1.6.0',
+]
+
+# [[linter]]
+# code = 'MYPY'
```
Comment (AllenTiTaiWang) [inline .lintrunner.toml:33]: Is MYPY still ignored even though it has linter.py?

File: .lintrunner/grep_linter.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
+# From PyTorch:
```
Comment (AllenTiTaiWang) [inline .lintrunner/grep_linter.py:1]: Does grep have settings in toml?

**Comment by AllenTiTaiWang** [comment]
What does new friends: newlines/grep do?

**Comment by gramalingam** [comment]
> I would like to make sure licensing and attribution is good for files under `.lintrunner/`

Do you mean from a legal perspective? Let me defer that to Aaron. I guess it should not be a problem.

File: .lintrunner/grep_linter.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Generic linter that greps for a pattern and optionally suggests replacements."""
```
Comment (justinchuby) [inline .lintrunner/grep_linter.py:79]: @AllenTiTaiWang right here

File: .lintrunner.toml (hunk 2 above)
Comment (justinchuby) [reply .lintrunner.toml:33]: Yes. We need to gradually enable mypy. Likely in subsequent prs.

File: .lintrunner/grep_linter.py (hunk 3 above)
Comment (justinchuby) [reply .lintrunner/grep_linter.py:1]: NOQA, SPACES and TABS. I think what they look for is apparent from the config. LMK if you need more info!

File: .lintrunner/newlines_linter.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""NEWLINE: Checks files to make sure there are no trailing newlines."""
```
Comment (justinchuby) [inline .lintrunner/newlines_linter.py:79]: @AllenTiTaiWang newline does this

**Comment by abock** [comment]
> > I would like to make sure licensing and attribution is good for files under `.lintrunner/`
> 
> Do you mean from a legal perspective? Let me defer that to Aaron. I guess it should not be a problem.

It's MIT licensed, so no issues here as long as the original license files, annotations, and copyrights are preserved.

However, is there a reason for including all the code in the repo? Can this be installed as a build dependency or should we submodule it?

I will do a deeper review tomorrow.

**Comment by justinchuby** [comment]
The original adapters are part of the pytorch repo. I guess we can create a package for them? For now they live in https://github.com/justinchuby/lintrunner-adapters

**Comment by justinchuby** [comment]
Alternatively we can create a new repo in the ms org and share it as a submodule across the various onnx projects

**Comment by AllenTiTaiWang** [comment]
> Alternatively we can create a new repo in the ms org and share it as a submodule across the various onnx projects

Feel like it makes more sense that each repo has its own adapters though. Or it could affect multiple repos, if one of the repo wanted a new formatting.

**Comment by justinchuby** [comment]
I agree. I feel we can allow max control by having the adapters local to each repo. If we see a more reusable pattern we can then move them to the same place.

File: MANIFEST.in [hunk 6]
```diff
@@ -1,10 +1,10 @@
-# include
-include *.rst
-include LICENSE
-include README.md
-include MANIFEST.in
-include requirements.txt
-
-# exclude from sdist
-exclude *.onnx
+# include
```
Comment (justinchuby) [inline MANIFEST.in:1]: Normalized line endings

**Review by gramalingam — APPROVED** [review]

**Pull request merged**